### DISCUSSION TITLE
Will Positron IDE eventually be made available to Posit Workbench as a web-based IDE?

### DISCUSSION THREAD

**Discussion by DaveGuenther** [posit-dev/positron#3695] [Q&A] [answered]
This would be nice to see along-side Code Server, Jupyter, and RStudio options.

**Comment by jthomasmock** [comment, marked as answer]
Thanks for your interest! 

Yes, we plan to have preview support for Positron sessions in Posit Workbench in 2024.

**Reply by AdaemmerP** [reply]
Will there be a free Positron server version as well? Similar to Rstudio server or Code server?

**Reply by jmcphers** [reply]
No, we don't currently plan to ship a free server version.

**Reply by jthomasmock** [reply]
As Jonathan said, we do not plan to release a stand-alone Positron server product. Instead, for now, we plan to offer a desktop-to-server connection via Remote SSH (or equivalent functionality), i.e., a single-user connection where a user themselves can maintain the capabilities without being forced to only work on the server. In this workflow, there would be a headless installation of Positron components on a remote server, but Positron is otherwise fully functional from the desktop with some optional computing happening in a server for efficient scalability or data access purposes.

**Reply by JosiahParry** [reply]
@jthomasmock i've had some people ask me for clarification on this here comment. 

It might be helpful to clarify in simpler words. As i understand it positron will essentially be a portal into an existing server (like vs code ssh tunneling). The experience people will get is essentially this: 

<img width="1011" alt="image" src="https://github.com/posit-dev/positron/assets/13534508/c097ac9c-a74a-466f-9c75-e48f74375ccb">

**Reply by jthomasmock** [reply]
Thanks, Josiah! Feel free to tag me if you get questions directly — I'm trying to keep up with posts across a few platforms.

Yes, that is roughly the right idea. Using desktop Positron, you can write code locally but execute it remotely, typically on a much more powerful server.

There are two extensions of interest for inspiration about the workflow:
- [VS Code Remote - SSH](https://code.visualstudio.com/docs/remote/ssh) - both proprietary to the VS Code marketplace and licensed to only to be used with VS Code. It is also not immediately compatible with Positron at technical level.
- [Open Remote - SSH](https://open-vsx.org/extension/jeanp413/open-remote-ssh) -- this is on OpenVSX and is not proprietary to VS Code. However, it is not immediately compatible with Positron, so we are planning to build our own implementation in a similar way.

**Reply by e-kotov** [reply]
Well, one can already run a Linux container on that beefy server with some lightweight GUI like Xfce and rdp/vnc, also install Positron in there, connect using RDP/VNC client.

**Reply by JosiahParry** [reply]
> Well, one can already run a Linux container on that beefy server with some lightweight GUI like Xfce and rdp/vnc, also install Positron in there, connect using RDP/VNC client.

I understand all of those letters separately. But together, I do not! 🙈 

Thanks @jthomasmock :)

**Reply by AdaemmerP** [reply]
Thanks for the clarification and the great IDE! 

However, "for now" does not preclude providing a server solution for the future, does it? My university provides Docker containers with software that students can run in the browser. So a web-browser-based solution to run Positron, such as code server, would be great.

**Reply by jthomasmock** [reply]
Thanks @AdaemmerP for the interest!

We already provide Posit Workbench for free to academic teaching groups. x-posting with: https://github.com/posit-dev/positron/issues/3730#issuecomment-2203786528

Once we have developed Positron server sessions in Workbench, we'd also like to offer two options:
* Free Posit Workbench, which would include RStudio, Positron, VS Code, JupyterLab, as well as authentication, support for Slurm/HPC and Kubernetes, etc
* Free Positron support for something like an existing JupyterHub deployment or a containerized model, or other arbitrary server hosting internally in a lightweight way.

Hosting and accessing persistent tools in a server is still a big ask for some groups, so we are also very motivated to do the Remote SSH-style workflow so that teams that can continue to work from a desktop can still get access to server-grade computing without having to do ALL of their work on a server.

I hope that is helpful!

**Reply by ZhimingYe** [reply]
Thanks @jthomasmock @jmcphers  for clarifying this matter. 
Another question is that I have been executing code in `RStudio Server` instead of `Jupyter Lab` or `VSCode`’s SSH service primarily because, regardless of the stability of my network connection or whether I close my laptop and then come back to refresh the browser, I can still get the output results on my webpage. (`Jupyter Lab` cannot achieve this; once the network connection is lost, it is very likely that the active notebook will not receive the background results. `VSCode` requires the use of terminal background software like `tmux` or `screen` to accomplish this, which is very inelegant.) Since `RStudio Server` has handled this very well in the past, I want to ask if `Positron` has any plans to optimize this process, because indeed `VSCode`’s SSH does not perform well in such scenarios.
Therefore, I was quite saddened to hear that you will no longer be offering services like “Positron Server” to individual users. So, I am presenting this need in the hopes of seeing it optimized in `Positron`. Additionally, do you have a specific timeline for the launch of your SSH service?

**Reply by juliasilge** [reply]
@ZhimingYe you can read (and subscribe, if you like) about plans for using the open source SSH tunneling extension in #2307.

**Reply by jrosell** [reply]
@jthomasmock Free Posit Workbench could be useful for SMB, but reading this thread, I wonder few things:

* Visual Studio Code Dev Containers feature is very nice for projects with people using mac, linux, windows (It uses Docker to replicate same IDE for multiple S.O). Could we have instructions to build a custom positron installation so we can be able have our same custom dev container in mac, linux, windows? 
* Another idea: Will it be possible to dockarize positron ide and be able to use it in the browser from mac, linux, windows?

I don't know if that is dreaming. Sorry if that's the case.

**Reply by jrosell** [reply]
> Well, one can already run a Linux container on that beefy server with some lightweight GUI like Xfce and rdp/vnc, also install Positron in there, connect using RDP/VNC client.

@e-kotov does it work in mac, linux and windows? Wouldn't be easier to use vs code dev containers or a browser?

**Reply by e-kotov** [reply]
> > Well, one can already run a Linux container on that beefy server with some lightweight GUI like Xfce and rdp/vnc, also install Positron in there, connect using RDP/VNC client.
> 
> @e-kotov does it work in mac, linux and windows? Wouldn't be easier to use vs code dev containers or a browser?

It works on a linux server (the container) with any VNC client on mac/linux/win. It would be easier to use dev containers, but at the moment you cannot install Positron in the container as you can the code-server. However you can install Positron as a GUI app in a container and connect via VNC as I suggested. But yeah, I am waiting for a dev container like experience.

**Reply by jthomasmock** [reply]
Howdy @jrosell thanks for the questions!

> @jthomasmock Free Posit Workbench could be useful for SMB, but reading this thread, I wonder few things:

We also have an SMB (Small Medium Business) policy for Posit Workbench, which is different from academic but includes discounting.

> * Visual Studio Code Dev Containers feature is very nice for projects with people using mac, linux, windows (It uses Docker to replicate same IDE for multiple S.O). Could we have instructions to build a custom positron installation so we can be able have our same custom dev container in mac, linux, windows?

The Dev Containers extension is proprietary and cannot be used with Positron or other variations of Code OSS. As a first step, we are currently focused on desktop-to-server remote connections (i.e., the Remote SSH equivalent) to allow users to access server-based compute or environments.

There are potential workflows where Positron is installed locally, and you use a Remote connection to a running container locally or on a remote server. As such, I don't want to speculate further about container support until we first address the remote connection tooling.

> * Another idea: Will it be possible to dockarize positron ide and be able to use it in the browser from mac, linux, windows?
> 
> I don't know if that is dreaming. Sorry if that's the case.

We don't expect to have native browser-based support for Positron outside of Posit Workbench at this time.
